### PR TITLE
Adding SMS notification templates to the registry on server startup and tenant creation

### PR DESCRIPTION
**Pull request by somindatommy**
## Proposed changes in this pull request
Have introduced new set of APIs to add default SMS notifications during the server startup and tenant creation.

Reelated to Issue: https://github.com/wso2/product-is/issues/7006

## When should this PR be merged
The PR https://github.com/wso2-extensions/identity-governance/pull/350 should be merged first and then the governance version needs to be updated in the current PR.

## Tests
Unit tests have provided.

### CODE REVIEW

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java [hunk 1]
```diff
@@ -571,6 +696,162 @@ public void addDefaultEmailTemplates(String tenantDomain) throws I18nEmailMgtExc
...
+        List<NotificationTemplate> notificationTemplates =
+                getDefaultNotificationTemplates(notificationChannel);
+        int numberOfAddedTemplates = 0;
+        try {
+            for (NotificationTemplate template : notificationTemplates) {
+                String displayName = template.getDisplayName();
+                String type = I18nEmailUtil.getNormalizedName(displayName);
+                String path = buildTemplateRootDirectoryPath(type, notificationChannel);
+
+            /*Check for existence of each category, since some template may have migrated from earlier version
```
Comment (NilukaSripalim) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java]: Use line comments

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java [hunk 2]
```diff
@@ -477,33 +510,107 @@ private String resolveNotificationChannel(String notificationChannel) {
         }
     }
 
+    /**
+     * Create a registry resource instance of the notification template.
+     *
+     * @param notificationTemplate Notification template
```
Comment (NilukaSripalim) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java]: Put full stop end of comment

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java [hunk 3]
```diff
@@ -571,6 +696,162 @@ public void addDefaultEmailTemplates(String tenantDomain) throws I18nEmailMgtExc
         }
     }
 
+    /**
+     * Add the default notification templates which matches the given notification channel to the respective tenants
+     * registry.
+     *
+     * @param notificationChannel Notification channel (Eg: SMS, EMAIL)
```
Comment (NilukaSripalim) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java]: Put the full stop, end of the comment

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java [hunk 4]
```diff
@@ -696,27 +1015,107 @@ private void validateTemplateType(String templateDisplayName, String tenantDomai
...
-            throw new I18nEmailMgtClientException("Locale code cannot be empty or null");
+        if (StringUtils.isBlank(displayName)) {
+            String errorCode =
+                    I18nEmailUtil.prependOperationScenarioToErrorCode(
+                            I18nMgtConstants.ErrorMessages.ERROR_CODE_EMPTY_TEMPLATE_NAME.getCode(),
+                            I18nMgtConstants.ErrorScenarios.EMAIL_TEMPLATE_MANAGER);
+            throw new NotificationTemplateManagerClientException(errorCode,
+                    I18nMgtConstants.ErrorMessages.ERROR_CODE_EMPTY_TEMPLATE_NAME.getMessage());
         }
+        /*Template name can contain only alphanumeric characters and spaces, it can't contain registry invalid
```
Comment (NilukaSripalim) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java]: Use line comments

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/util/I18nEmailUtil.java [hunk 5]
```diff
@@ -271,5 +273,27 @@ public static Collection createTemplateType(String normalizedTemplateName, Strin
         return collection;
     }
 
+    /**
+     * Prepend the operation scenario to the existing exception error code.
+     * (Eg: USR-20045)
+     *
+     * @param exceptionErrorCode Existing error code.
+     * @param scenario           Operation scenario
+     * @return New error code with the scenario prepended (NOTE: Return an empty String if the provided error code is
```
Comment (NilukaSripalim) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/util/I18nEmailUtil.java]: Put the full stop, end of the comment

File: features/org.wso2.carbon.email.mgt.server.feature/resources/sms-templates-admin-config.xml [hunk 6]
```diff
@@ -0,0 +1,41 @@
+<!--
+ ~ Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (NilukaSripalim) [inline features/org.wso2.carbon.email.mgt.server.feature/resources/sms-templates-admin-config.xml]: Use standard wso2 copyright. (~ instead *)

File: components/email-mgt/org.wso2.carbon.email.mgt/src/test/resources/sms/sms-templates-admin-config.xml [hunk 7]
```diff
@@ -0,0 +1,41 @@
+<!--
+ ~ Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (NilukaSripalim) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/test/resources/sms/sms-templates-admin-config.xml]: Use standard wso2 copyright. (~ instead *)

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java (hunk 3 above)
Comment (ShanChathusanda93) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java]: There are several other places with missing full stops. Please add them accordingly.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java [hunk 8]
```diff
@@ -477,33 +510,107 @@ private String resolveNotificationChannel(String notificationChannel) {
         }
     }
 
+    /**
+     * Create a registry resource instance of the notification template.
+     *
+     * @param notificationTemplate Notification template
+     * @return Resource
+     * @throws NotificationTemplateManagerServerException Error while creating the resource
```
Comment (vihanga-liyanage) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java]: @throws NotificationTemplateManagerServerException If an error occurred while creating the resource.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/internal/I18nMgtServiceComponent.java [hunk 9]
```diff
@@ -86,8 +88,9 @@ protected void activate(ComponentContext context) {
                 log.error("I18n Management - TenantMgtListener could not be registered");
             }
 
-            // Load default email templates.
+            // Load default notification templates.
             loadDefaultEmailTemplates();
+            loadDefaultSMSNotificationTemplates();
```
Comment (vihanga-liyanage) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/internal/I18nMgtServiceComponent.java]: Shouldn't this be ```loadDefaultSMSTemplates()```?

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java (hunk 1 above)
Comment (somindatommy) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java]: This is a multiline comment. Therefore, multiline comments will be used.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java (hunk 2 above)
Comment (somindatommy) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java]: `.` has been added to the method level comment. We do not add `.` at the end of param descriptions.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java (hunk 3 above)
Comment (somindatommy) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java]: `. `has been added to the method level comment. We do not add `.` at the end of param descriptions.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/util/I18nEmailUtil.java (hunk 5 above)
Comment (somindatommy) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/util/I18nEmailUtil.java]: `.` has been added to the method level comment. We do not add `.` at the end of param descriptions.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java (hunk 4 above)
Comment (somindatommy) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java]: This is a multiline comment. Therefore, using line comments is wrong.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/constants/I18nMgtConstants.java [hunk 10]
```diff
@@ -61,4 +64,62 @@ private ErrorMsg() {
...
+        // ETM - EMAIL TEMPLATE MANAGER
+        public static final String EMAIL_TEMPLATE_MANAGER = "ETM";
+    }
+
+    /**
+     * Enum which contains error codes and corresponding error messages.
+     */
+    public enum ErrorMessages {
+
+        ERROR_CODE_NULL_TEMPLATE_OBJECT("60001", "Notification template object cannot be null."),
```
Comment (IsuraD) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/constants/I18nMgtConstants.java]: Notification template  is not provided

Comment (somindatommy) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/constants/I18nMgtConstants.java]: Fixed: 008a5546baef7725f1a4c04554942b4db91f41f8

**Pull request merged**